### ISSUE TITLE
Installing requirements when using virtualenv

### ISSUE THREAD

**Issue by DominikRafacz** [bodkan/slendr#84]
Hi! Thanks for the package! I can see a lot of good work put into it!

I have noticed one issue with the installation of the requirements when using _virtualenv_. You are using _conda_ by default suggested installation, so this problem could go under your radar.

Steps to recreate:

1. Setup a new _virtualenv_: `reticulate::virtualenv_create("./venv")`
2. Select newly created _virtualenv_ for _slendr_ : `slendr::setup_env("./venv")`
```
Successfully connected to the specified Python virtual environment './venv'
Python module msprime is missing in the environment. Install? 

1: No
2: Yes
```
3. Select "2"
```
Selection: 2
Using virtual environment "./venv" ...
ERROR: Invalid requirement: 'msprime=1.1.0'
Hint: = is not a valid operator. Did you mean == ?
Error: Error installing package(s): "\"msprime=1.1.0\""
```

This problem also appears when trying to install any other python package via this super handy `setup_env`. I've digged into the code and I've noticed that, indeed, you're using single `=`, while `requirements.txt` for _virtualenv_ requires double `==`. 

I'm not using _conda_ very often, I prefer _virtualenv_ (that's how I've stumbled across it), but after quick googling I can tell that `=` works for _conda_. I've also checked it -- it works.

**Comment by DominikRafacz** [comment]
The simple solution for it would be to separate _conda_ and _virtualenv_ in the `if` code checking for dependencies, or replace `=` with `==` (but actually, again -- I'm not familiar enough with _conda_syntax).

I can also make a quick pull request, if you find it suitable.

**Comment by bodkan** [comment]
Hey, thanks for opening the issue!

(Also, LOL at the first to _slendr_ issues being actually Python issues. I had nightmares this will happen.)

I'll be honest. I've been considering making `setup_env()` only create the standalone miniconda environment with reticulate. Seeing these issues makes me even more convinced this is the right way to do this.

Even without the usual problems that [another brave user](https://github.com/bodkan/slendr/issues/83) just reported, I wonder if it's too dangerous to let users link slendr to any version of msprime/tskit/pyslim combination.

The only reason `setup_env()` exists is to make it easier for people to link R to a dedicated slendr env without having to worry about the details. My intention was that whatever someone uses to manage environments for their Python work, slendr has a right to do it its own way. :) Even just to save me headaches in the future. 🤔 

Do you have thoughts on this?

In other words, this:

> I'm not using conda very often, I prefer virtualenv

shouldn't matter for working with slendr because whatever slendr does under the hood should be a low level detail.

[And I'm saying all this as someone who doesn't use conda at all and only uses venv. The conda env that `setup_env()` automatically creates is the only conda instance on my machine.]

**Comment by bodkan** [comment]
OK, having thought about it more and looking at the monstrosity that `setup_env()` has become during development, I decided to dedicate `setup_env()` purely to setting up the automatic miniconda environment. The target audience of this function has always been a user who doesn't really care about what Python env is being used by slendr, as long as it's doing its job. This also seems what the reticulate team seems to be recommending (as a default way of automatic Python setup across platforms).

For anyone else (users with custom venv/virtualenv/pyvenv/...), the good old `reticulate::use_virtualenv` (your case) or `reticulate::use_condaenv` (for conda users who want to use specific environment) is always there - [link](https://rstudio.github.io/reticulate/reference/use_python.html). Note that in your situation, this is exacly what `setup_env()` was doing, so it didn't even save typing.

My reasoning behind this decision:

`setup_env()` offered a complex way to interactively install individual Python modules which is something that anyone who wants to create their own environments can do on their own much more robustly with standard Python tools than I ever could with any kind of R interface. It was very kind of you to say "super handy `setup_env`" but I worry that I made the wrong decision with that function.

The number of scenarios in which things could break by this step-by-step Python module installation process in R (who knows what kinds of other modules could users have installed there which would cause dependency conflicts) is not worth it. I'm afraid I have no capacity to give support in these situations (not really being a Python user myself).

These advanced users can turn off the annoying message announcing the possibility to use `setup_env()` for automatic miniconda setup by setting `options(slendr.custom_env = TRUE)`.

---

Could you verify that you can create your own virtualenv (by whatever means) and link it with R using `reticulate::use_virtualenv("~/myenv")`? This is what I was doing during the 6 months of slendr work in the time `setup_env()` didn't yet exist (I wrote that primarily for newbies and complete non-Python users).

You seem to be familiar with reticulate, but I will link to the docs [just in case](https://rstudio.github.io/reticulate/reference/use_python.html).

**Comment by DominikRafacz** [comment]
Yes, I can create a new _virtualenv_, call the `use_virtualenv` function and have my R session linked to it. I've installed the required packages manually from the terminal. `slendr` works correctly after those steps, I've checked by calling `check_env()`. 

I've alread done that before, using _virtualenvs_ and _reticulate_ is not a problem for me. I've opened this issue because I think that `setup_env` function is a cool, quick way of setting things up. Having one function which will install the dependencies for you is a really convenient thing. I'd love to be able to use it even if I create virtualenv manually, as I prefer. This is not a big deal for me, just a convenience.

If I were you, I'd simply fix this installation of packages issue with separate code for installing python dependencies for _virtualenv_ and _conda_ environments. I'd also probably browse through other packages too see how they do it.

I'm familiar with `reticulate`, that's true. I'm not familiar, however, with using it as a dependency of a regular package. Having external dependencies in other language seems quite problematic and I can feel your struggle (I've got only experience with calling C and C++ libraries, which case is more established). I don't know what are the standards (are there any?) for python so you can just discard my suggestion and simply inform users, that if they provide their custom python interpreter in custom environment (except for _conda_), they need to install the packages manually.

**Comment by bodkan** [comment]
> Yes, I can create a new virtualenv, call the use_virtualenv function and have my R session linked to it. I've installed the required packages manually from the terminal. slendr works correctly after those steps, I've checked by calling check_env().

Excellent! 👏  Thank you so much for confirming that this works! That's a massive relief! 😅 

> I've opened this issue because I think that setup_env function is a cool, quick way of setting things up.

Yes, this is why I wrote it, originally. But looking over the code again made me realize that if the user calls `setup_env("~/env-name")` and `env-name` is somehow corrupted (only some dependencies, partially installed, or any number of issues that can occur with venv/virtualenv/conda), I would be digging a really deep hole for myself as a developer who needs to support this.

> f I were you, I'd simply fix this installation of packages issue with separate code for installing python dependencies for virtualenv and conda environments.

Yes, I will add this info to the current [installation guide](https://www.slendr.net/articles/vignette-00-installation.html#python). Making sure advanced Python users are aware of the fact that `reticulate:: use_virtualenv` can be used to link to a custom environment.

That said, from a practical perspective, I don't see much of a reason to do this instead of using `setup_env()`. After all, the only trio of msprime/tskit/pyslim I can support is the one which has the same versions of the packages contained in what would be in the automated `setup_env()` as well.

Also, if the automatic slendr conda env _is_ present, `library(slendr)` will activate it without any need for explicit `reticulate::use_virtualenv` or `reticulate::use_condaenv`, which seems like a bigger convenience, actually.

>  I'd also probably browse through other packages too see how they do it.

I modelled `setup_env` behavior (which auto-installs and configures a dedicated, isolated mimoconda env) after what reticulate seems to be recommending and also what they seem to be doing with their tensorflow and keras R-Python packages.

I was against this at first (I don't really like these automated solutions) but when I actually had to roll up my sleeves and implement a reasonably robust way to link slendr to Python, I could see why this is the easiest way (+ there's the issue that it's the only reasonable way to support on Windows).

**Comment by bodkan** [comment]
Anyway, sorry for this long rant! I will keep it as a reference for potential future issues on a similar topic.

Many thanks for reporting this problem and for your quick response! I had a bit of a panic there for a second and I appreciate the help.

Thanks for checking out the package and please, don't hesitate to report more issues!

Also, a huge **"yes please!**" for future offers of pull requests in case you run into more problems that went under my radar (of which I'm sure there are many)! Many thanks for the offer. ❤️  The _only_ reason I decided to go against this (in this case) was the nightmarish vision about having (potentially) a partial responsibility for custom "non-standard" Python environments. That's why decided to simplify the `setup_env()` code drastically. Your suggestion was otherwise perfectly reasonable.

**Comment by bodkan** [comment]
The changes discussed above are now implemented in https://github.com/bodkan/slendr/commit/30f24b9 and described in the [changelog](https://github.com/bodkan/slendr/blob/main/NEWS.md) (currently listed under "Development version").

**Comment by DominikRafacz** [comment]
Okay, thanks for the detailed response and explaination! I think this settles the issue, I'm closing it.
